### ISSUE TITLE
Does sja1105-tool send the configuration directly to the sja1105 chip?

### ISSUE THREAD

**Issue by liu-jeff** [openil/sja1105-tool#58]
Hello, I want to know that sja1105-tool sends configuration through Ioctl function. Is there a receiving function in kernel ?

**Comment by vladimiroltean** [comment]
sja1105-tool uses the generic SPI driver (spidev), which is both its best feature and its worst weakness.
There also exists a native kernel driver which makes the switch look much more similar to Felix from LS1028A (net devices for switch ports, PHY management, VLAN and FDB manipulation with standard iproute2 tools, etc etc).
Why ask?

**Comment by liu-jeff** [comment]
> Hello, I want to know that sja1105-tool sends configuration through Ioctl function. Is there a receiving function in kernel mode?



> sja1105-tool uses the generic SPI driver (spidev), which is both its best feature and its worst weakness.
> There also exists a native kernel driver which makes the switch look much more similar to Felix from LS1028A (net devices for switch ports, PHY management, VLAN and FDB manipulation with standard iproute2 tools, etc etc).
> Why ask?
I was reading the code of openil, and only found the interface of Netlink in the kernel, but not the interface of ioctl.

**Comment by vladimiroltean** [comment]
This is the kernel driver used by sja1105-tool:
https://github.com/openil/linux/blob/master/drivers/spi/spidev.c

**Comment by liu-jeff** [comment]
> This is the kernel driver used by sja1105-tool:
> https://github.com/openil/linux/blob/master/drivers/spi/spidev.c

Thank you very much, so is the qbv function in LS1021ATSN implemented in the kernel or sja1105?

**Comment by vladimiroltean** [comment]
It is implemented in hardware.
The configuration is done from sja1105-tool.
In the native kernel driver I was mentioning, it is done through the standard tc-taprio offload command (still not tsntool).

**Comment by liu-jeff** [comment]
> In the native kernel driver I was mentioning, it is done through the standard tc-taprio offload command (still not tsntool).
So for LS1021ATSN, if it forwards the data packet, will it not go to the kernel? Is this packet not extracted in the kernel?

**Comment by liu-jeff** [comment]
> It is implemented in hardware.
> The configuration is done from sja1105-tool.
> In the native kernel driver I was mentioning, it is done through the standard tc-taprio offload command (still not tsntool).

In fact, I just want to extract the data packets forwarded by the switch in the kernel and perform some operations on them.

**Comment by liu-jeff** [comment]
> It is implemented in hardware.
> The configuration is done from sja1105-tool.
> In the native kernel driver I was mentioning, it is done through the standard tc-taprio offload command (still not tsntool).
the standard tc-taprio offload command can achieve qbv?

**Comment by vladimiroltean** [comment]
Yes, see [the documentation](https://www.kernel.org/doc/html/latest/networking/dsa/sja1105.html#time-aware-scheduling) for it.

**Comment by liu-jeff** [comment]
> Yes, see [the documentation](https://www.kernel.org/doc/html/latest/networking/dsa/sja1105.html#time-aware-scheduling) for it.

But can your native kernel code be ported to openil?
